spec/models/api/v1/student_registration_spec.rb

```diff
@@ -29,4 +29,33 @@
     its(:student) { should be_valid }
   end
 
-end      
\ No newline at end of file
+  describe "#save" do
+    context "given duplicate user names" do
+      it "creates unique logins" do
+
+        @count = User.all.size
+
+        reg1 = API::V1::StudentRegistration.new(params)
+        reg1.save
+
+        #
+        # Set the login attribute to be the login we know was created
+        # by the previous save. Note this requires the test to have
+        # some knowledge of the implementation. Should the implementation
+        # change (or even the interface e.g. attribute access) then this
+        # test might no longer be valid.
+        #
+        # See bug PT #109387108
+        #
+        reg2 = API::V1::StudentRegistration.new(params)
+        reg2.login = 'sdoe'
+        reg2.save
+
+        expect(User.all.size).to eql(@count + 2)
```

Comment:
I'm not sure if that fully answers all of your questions, but it explains the behavior of the test, which relies on, or exploits, this stateful nature of the "caching" (or re-use) of the @user object in the base class.